helpers/DeployHelper.ts

```diff
@@ -40,72 +41,84 @@ export class DeployHelper {
-
-    let g16Verifier;
-    if (
-      ["Groth16VerifierStateTransition", "Groth16VerifierStub"].includes(g16VerifierContractName)
-    ) {
-      g16Verifier = await ethers.deployContract(g16VerifierContractName);
-    } else {
-      throw new Error("invalid verifierContractName");
-    }
-    await g16Verifier.waitForDeployment();
-    this.log(
-      `${g16VerifierContractName} contract deployed to address ${await g16Verifier.getAddress()} from ${await owner.getAddress()}`,
+    const [poseidon1Elements, poseidon2Elements, poseidon3Elements] = await deployPoseidons(
+      [1, 2, 3],
+      deployStrategy,
     );
 
-    if (poseidonContracts.length === 0 || poseidonContracts.length !== 3) {
-      this.log("deploying poseidons...");
-
-      const tx = await g16Verifier.deploymentTransaction();
-      await waitNotToInterfereWithHardhatIgnition(tx);
-
-      const [poseidon1Elements, poseidon2Elements, poseidon3Elements] = await deployPoseidons(
-        [1, 2, 3],
-        deployStrategy,
-      );
-      poseidonContracts.push(poseidon1Elements, poseidon2Elements, poseidon3Elements);
-    }
-
-    const poseidon1Elements = poseidonContracts[0];
-    const poseidon2Elements = poseidonContracts[1];
-    const poseidon3Elements = poseidonContracts[2];
-
-    this.log("deploying SmtLib...");
     const smtLib = await this.deploySmtLib(
       await poseidon2Elements.getAddress(),
       await poseidon3Elements.getAddress(),
       "SmtLib",
       deployStrategy,
     );
 
+    const groth16VerifierStateTransition = await this.deployGroth16VerifierStateTransition(
+      g16VerifierContractName,
+      deployStrategy,
+    );
+
+    const { state, stateLib, stateCrossChainLib, crossChainProofValidator, defaultIdType } =
+      await this.deployState(
+        supportedIdTypes,
+        deployStrategy,
+        await smtLib.getAddress(),
+        await poseidon1Elements.getAddress(),
+        await groth16VerifierStateTransition.getAddress(),
+      );
+
+    return {
+      state,
+      stateLib,
+      stateCrossChainLib,
+      crossChainProofValidator,
+      defaultIdType,
+      smtLib,
+      poseidon1: poseidon1Elements,
+      poseidon2: poseidon2Elements,
+      poseidon3: poseidon3Elements,
+      groth16verifier: groth16VerifierStateTransition,
+    };
+  }
+
+  async deployState(
+    supportedIdTypes: string[] = [],
+    deployStrategy: "basic" | "create2" = "basic",
+    smtLibAddress: string,
+    poseidon1Address: string,
+    groth16verifierAddress: string,
+  ): Promise<{
+    state: Contract;
+    stateLib: Contract;
+    stateCrossChainLib: Contract;
+    crossChainProofValidator: Contract;
+    defaultIdType;
+  }> {
+    this.log("======== State: deploy started ========");
+
+    const { defaultIdType, chainId } = await this.getDefaultIdType();
+    this.log(`found defaultIdType ${defaultIdType} for chainId ${chainId}`);
+
+    const owner = this.signers[0];
+
     this.log("deploying StateLib...");
```

Comment:
Can you write the address of the stateLib, stateCrossChainLib, CrossChainProofValidator contract to some kind of a temporary log (only if not in local network) and if the deployment process breaks, check and ask a user if need to redeploy. Delete the temporary log at the end of the script run maybe.
The idea is to limit the deploy gas usage in case if the deployments breaks in the middle.

Reply:
Yes, I'll do it.